Add caching to ci

ryanio opened the pull request:
## Description

Thanks to @jjangga0214 in https://github.com/ethereum/web3.js/pull/3547, this PR adds to his work to add caching to our ci.

This PR also:

* Normalizes deps to use lerna [`--hoist`](https://github.com/lerna/lerna/blob/master/doc/hoist.md) option to reduce the size of node_modules cache
* Sets typescript devDep version from `latest` to `^3.9.5`
* Removes `puppeteer`
  * I don't think it was being used and made the dev ex on `npm install` really annoying by downloading a 110mb package of chrome unless you set `PUPPETEER_SKIP_CHROMIUM_DOWNLOAD=true`
* Updates package-locks to latest

## Type of change

<!-- Please delete options that are not relevant. -->

- [x] ci enhancements

jjangga0214 committed:
ci(build.yml): ignore paths

jjangga0214 committed:
build(deps): update package-lock.json files

jjangga0214 committed:
change to `npm ci`

jjangga0214 committed:
ci(build.yml): cache npm by lock file

jjangga0214 committed:
ci(build.yml): refactor

remove empty line

ryanio committed:
remove puppeteer

ryanio committed:
fix caching

ryanio committed:
update package-locks

ryanio committed:
remove install from windows ci (doesn't do anything)

ryanio committed:
CI is default true

ryanio committed:
skip install on cache hit

ryanio commented on .github/workflows/build.yml:
```diff
@@ -5,8 +5,22 @@ on:
+            - "LICENSE"
     pull_request:
         types: [opened, reopened, synchronize]
+        paths-ignore:
```
I like the idea of the paths-ignore added here to be more efficient with the ci, but I'm not sure what happens if the checks are required and they don't get run - would the PR be blocked from merging?

@jjangga0214 do you know?

ryanio replied:
I tried this on my fork and the status checks are still required thus the PR cannot be merged in, so I will remove these `paths-ignore` for now, although it is a nice idea.

ryanio committed:
pin `typescript` devDep to `^3.9.5`

ryanio committed:
update cache paths based on this example: https://github.com/actions/cache/blob/master/examples.md#node---lerna

ryanio committed:
update `package.json`s to satisfy lerna hoisting

GregTheGreek commented on test/utils.toTwosComplement.js:
```diff
@@ -7,7 +7,7 @@ var utils = require('../packages/web3-utils/');
 
 var tests = [
-    { value: 1, expected: utils.padLeft(new BigNumber(1).round().toString(16), 64) },
+    { value: 1, expected: utils.padLeft(new BigNumber(1).integerValue().toString(16), 64) },
```
Are we introducing some potential breaking features by bumpng to `bignumber.js: 9.0.0` ?

GregTheGreek replied:
ah, i see it was being used previously at two different versions

ryanio replied:
`bignumber.js` in `web3-core` was already on `9.0.0`, this PR just bumps the one in the root node_modules so it's normalized for hoisting. I think it only affects `bignumber.js` as used in the test suite and the above (round => integerValue) is the only breaking change I stumbled upon, but I will keep an eye out for other side effects.

cgewecke commented on karma.conf.js:
```diff
@@ -1,9 +1,5 @@
-if (!process.env.GITHUB_ACTION) {
-    process.env.CHROME_BIN = require('puppeteer').executablePath();
-}
-
```
Tiny detail but perhaps we can mark the browser tests `CI only` in [TESTING.md][1] as well. Think this change might make running those locally fail. 

[1]: https://github.com/ethereum/web3.js/blob/1.x/TESTING.md#required-tests

ryanio replied:
@cgewecke thx, I think I have it working locally now with https://github.com/ethereum/web3.js/pull/3609/commits/c5430edbfc3d79a932b43422d7a589a7398f7c50

ryanio committed:
simplify karma config

ryanio committed:
update ci `on` logic:
  remove paths-ignore since PR checks would still be required
  run only on PRs for `1.x` base branch

ryanio committed:
add GITHUB_TOKEN for bundlesize

cgewecke reviewed (approved):
This all looks good to me! Nice work. 💯  💯
